src/main/java/jenkins/plugins/git/GitToolChooser.java

```diff
@@ -106,6 +136,150 @@ private void decideAndUseAPI(String remoteName, Item context, String credentials
         }
     }
 
+    /* Git repository URLs frequently end with the ".git" suffix.
+     * However, many repositories (especially https) do not require the ".git" suffix.
+     *
+     * Add remoteURL with the ".git" suffix and without the ".git" suffix to the
+     * list of alternatives.
+     */
+    private void addSuffixVariants(@NonNull String remoteURL, @NonNull Set<String> alternatives) {
+        alternatives.add(remoteURL);
+        String suffix = ".git";
+        if (remoteURL.endsWith(suffix)) {
+            alternatives.add(remoteURL.substring(0, remoteURL.length() - suffix.length()));
+        } else {
+            alternatives.add(remoteURL + suffix);
+        }
+    }
+
+    /* Git repository URLs frequently end with the ".git" suffix.
+     * However, many repositories (especially https) do not require the ".git" suffix.
+     *
+     * Add remoteURL with the ".git" suffix if not present
+     */
+    private String addSuffixVariant(@NonNull String canonicalURL) {
+        String suffix = ".git";
+        if (!canonicalURL.endsWith(suffix)) {
+            canonicalURL = canonicalURL + suffix;
+        }
+        return canonicalURL;
+    }
+
+    /* Protocol patterns to extract hostname and path from typical repository URLs */
+    private static Pattern gitProtocolPattern = Pattern.compile("^git://([^/]+)/(.+?)/*$");
+    private static Pattern httpProtocolPattern = Pattern.compile("^https?://([^/]+)/(.+?)/*$");
+    private static Pattern sshAltProtocolPattern = Pattern.compile("^[\\w]+@(.+):(.+?)/*$");
+    private static Pattern sshProtocolPattern = Pattern.compile("^ssh://[\\w]+@([^/]+)/(.+?)/*$");
+
+    /* Return a list of alternate remote URL's based on permutations of remoteURL.
+     * Varies the protocol (https, git, ssh) and the suffix of the repository URL.
+     * Package protected for testing
+     */
+    /* package */ @NonNull String convertToCanonicalURL(String remoteURL) {
+        if (remoteURL == null || remoteURL.isEmpty()) {
+            LOGGER.log(Level.FINE, "Null or empty remote URL not cached");
+            return ""; // return an empty string
+        }
+
+        Pattern [] protocolPatterns = {
+                gitProtocolPattern,
+                sshAltProtocolPattern,
+                sshProtocolPattern,
+        };
+
+        String matcherReplacement = "https://$1/$2";
+        /* For each matching protocol, convert alternatives to canonical form by https replacement */
+        remoteURL = addSuffixVariant(remoteURL);
+        String canonicalURL = remoteURL;
+        if (httpProtocolPattern.matcher(remoteURL).matches()) {
+            canonicalURL = remoteURL;
+        } else {
+            for (Pattern protocolPattern: protocolPatterns) {
+                Matcher protocolMatcher = protocolPattern.matcher(remoteURL);
+                if (protocolMatcher.matches()) {
+                    canonicalURL = protocolMatcher.replaceAll(matcherReplacement);
```

Comment:
We can end the iteration on the first match.  No need to continue searching once we have a match.
```suggestion
                    canonicalURL = protocolMatcher.replaceAll(matcherReplacement);
                    break;
```

Reply:
Great catch, thanks.

Reply:
Is there a specific reason you're choosing to continue through all patterns after detecting a match?